mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/ClientState.scala

```diff
@@ -0,0 +1,526 @@
+          connectData,
+          Vector.empty,
+          queue,
+          data.consumerPacketRouter,
+          data.producerPacketRouter,
+          data.subscriberPacketRouter,
+          data.unsubscriberPacketRouter,
+          data.settings
+        )
+      )
+  }
+
+  def disconnect(context: ActorContext[Event],
+                 connectFlags: ConnectFlags,
+                 remote: SourceQueueWithComplete[ForwardConnectCommand],
+                 data: Data): Behavior[Event] = {
+    if (connectFlags.contains(ConnectFlags.CleanSession))
+      context.children.foreach(context.stop)
+
+    remote.complete()
+
+    disconnected(
+      Uninitialized(data.consumerPacketRouter,
+                    data.producerPacketRouter,
+                    data.subscriberPacketRouter,
+                    data.unsubscriberPacketRouter,
+                    data.settings)
+    )
+  }
+
+  def serverConnect(data: ConnectReceived): Behavior[Event] = Behaviors.withTimers { timer =>
+    timer.startSingleTimer("receive-connack", ReceiveConnAckTimeout, data.settings.receiveConnAckTimeout)
+
+    Behaviors
+      .receivePartial[Event] {
+        case (context, ConnAckReceivedFromRemote(connAck, local))
+            if connAck.returnCode.contains(ConnAckReturnCode.ConnectionAccepted) =>
+          local ! ForwardConnAck(data.connectData)
+          if (data.connect.connectFlags.contains(ConnectFlags.CleanSession))
+            context.children.foreach(context.stop)
+          data.stash.foreach(context.self.tell)
+          serverConnected(
+            ConnAckReceived(
+              data.connect.connectFlags,
+              data.connect.keepAlive,
+              pendingPingResp = false,
+              Vector.empty,
+              data.remote,
+              data.consumerPacketRouter,
+              data.producerPacketRouter,
+              data.subscriberPacketRouter,
+              data.unsubscriberPacketRouter,
+              data.settings
+            )
+          )
+        case (context, ConnAckReceivedFromRemote(_, local)) =>
+          local ! ForwardConnAck(data.connectData)
+          disconnect(context, data.connect.connectFlags, data.remote, data)
+        case (context, ReceiveConnAckTimeout) =>
+          disconnect(context, data.connect.connectFlags, data.remote, data)
+        case (_, e) if data.stash.size < data.settings.maxClientConnectionStashSize =>
+          serverConnect(data.copy(stash = data.stash :+ e))
+      }
+      .receiveSignal {
+        case (_, PostStop) =>
+          data.remote.complete()
+          Behaviors.same
+      }
+
+  }
+
+  def serverConnected(data: ConnAckReceived): Behavior[Event] = Behaviors.withTimers { timer =>
+    if (data.keepAlive.toMillis > 0)
+      timer.startSingleTimer("send-pingreq", SendPingReqTimeout, data.keepAlive)
+    else
+      data.remote.complete() // We'll never be sending pings so free up the command channel for other things
+
+    Behaviors
+      .receivePartial[Event] {
```

Comment:
Is the reason that you use `receivePartial` instead of `receive` that you don't care about several of the messages in certain FSM states? Otherwise the total function has the nice property that the compiler will help with exhaustiveness checks of sealed traits.

Reply:
Yes, this is deliberate and as illustrated by your documentation: https://doc.akka.io/docs/akka/2.5.17/typed/fsm.html. Would you personally list each event, given that there can be many?

Reply:
it's perfectly fine, I was just curious. If it was just about two or three messages I would probably list them to keep the exhaustiveness checks, but I fully understand that it can be too much boilerplate if many